print("Hello ML! 🚀")

# Breast Cancer Classification & Model Evaluation

# 1. Project Overview

## This project demonstrates binary classification using the Breast Cancer Wisconsin dataset. Logistic Regression and Random Forest models are trained and evaluated using accuracy, precision, recall, F1-score, ROC curve, and AUC.

## 2. Import Libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


# 3. Load Dataset

In [8]:
from sklearn.datasets import load_breast_cancer

data = load_breast_cancer()

print("Dataset loaded successfully!")

Dataset loaded successfully!


In [12]:
import pandas as pd

df = pd.DataFrame(
    data.data,
    columns=data.feature_names
)

df["target"] = data.target

df.head()

,mean radius,mean texture,mean perimeter,mean area,mean smoothness,mean compactness,mean concavity,mean concave points,mean symmetry,mean fractal dimension,...,worst texture,worst perimeter,worst area,worst smoothness,worst compactness,worst concavity,worst concave points,worst symmetry,worst fractal dimension,target
0,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.3001,0.14710,0.2419,0.07871,...,17.33,184.60,2019.0,0.1622,0.6656,0.7119,0.2654,0.4601,0.11890,0
1,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.0869,0.07017,0.1812,0.05667,...,23.41,158.80,1956.0,0.1238,0.1866,0.2416,0.1860,0.2750,0.08902,0
2,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.1974,0.12790,0.2069,0.05999,...,25.53,152.50,1709.0,0.1444,0.4245,0.4504,0.2430,0.3613,0.08758,0
3,11.42,20.38,77.58,386.1,0.14250,0.28390,0.2414,0.10520,0.2597,0.09744,...,26.50,98.87,567.7,0.2098,0.8663,0.6869,0.2575,0.6638,0.17300,0
4,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.1980,0.10430,0.1809,0.05883,...,16.67,152.20,1575.0,0.1374,0.2050,0.4000,0.1625,0.2364,0.07678,0


In [10]:

print("Number of samples:", data.data.shape[0])
print("Number of features:", data.data.shape[1])
print("Feature names:")
print(data.feature_names)

Number of samples: 569
Number of features: 30
Feature names:
['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


In [11]:
print("Dataset shape:", data.data.shape)

print("Target shape:", data.target.shape)

print("Number of features:", len(data.feature_names))

print("Target classes:", data.target_names)

Dataset shape: (569, 30)
Target shape: (569,)
Number of features: 30
Target classes: ['malignant' 'benign']


## 4. EDA

In [ ]:
df.shape

In [ ]:
df.info()

In [ ]:
df["target"].value_counts()

In [ ]:
df.isnull().sum()

In [ ]:
df.duplicated().sum()

In [ ]:
df.describe()

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

sns.countplot(x="target", data=df)

plt.title("Distribution of Target Classes")
plt.xlabel("Class")
plt.ylabel("Number of Samples")
plt.show()

In [ ]:
df["diagnosis"] = df["target"].map({
    0: "Malignant",
    1: "Benign"
})

sns.countplot(x="diagnosis", data=df)

plt.title("Malignant vs Benign Samples")
plt.xlabel("Diagnosis")
plt.ylabel("Number of Samples")
plt.show()

In [ ]:
df["target"].value_counts()

In [ ]:
df["target"].value_counts(normalize=True) * 100

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns


sns.countplot(x="target", data=df)

plt.title("Target Class Distribution")
plt.xlabel("Class (0 = Malignant, 1 = Benign)")
plt.ylabel("Number of Samples")
plt.show()

In [ ]:
df.columns

In [ ]:
df[["mean radius", "mean texture", "mean perimeter", "mean area"]].describe()

## 5. Data Preparation

X = df.drop(columns=["target", "diagnosis"])
y = df["target"]

print("X shape:", X.shape)
print("y shape:", y.shape)

## 6.Train-Test Split

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Training samples:", X_train.shape[0])
print("Testing samples:", X_test.shape[0])

## 7.Feature Scaling

In [ ]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

In [ ]:
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

## 8.Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=1000)

model.fit(X_train_scaled, y_train)

print("Model trained successfully!")

In [ ]:
y_pred = model.predict(X_test_scaled)

print(y_pred[:20])

## 9.Confusion Matrix

In [ ]:
plt.figure(figsize=(6, 5))

sns.heatmap(
    cm,
    annot=True,
    fmt="d",
    cmap="Blues",
    xticklabels=["Malignant", "Benign"],
    yticklabels=["Malignant", "Benign"]
)

plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.title("Confusion Matrix")

plt.savefig(
    "results/confusion_matrix.png",
    dpi=300,
    bbox_inches="tight"
)

plt.show()

In [ ]:
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred)

print(cm)

In [ ]:
plt.figure(figsize=(6, 5))

sns.heatmap(
    cm,
    annot=True,
    fmt="d",
    cmap="Blues",
    xticklabels=["Malignant", "Benign"],
    yticklabels=["Malignant", "Benign"]
)

plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.title("Confusion Matrix")
plt.show()

## 10.Classification Metrics

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Accuracy :", accuracy)
print("Precision:", precision)
print("Recall   :", recall)
print("F1 Score :", f1)

In [ ]:
print(f"Accuracy : {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall   : {recall:.4f}")
print(f"F1 Score : {f1:.4f}")

In [ ]:
y_pred

In [ ]:
y_prob = model.predict_proba(X_test_scaled)[:, 1]

In [ ]:
y_prob = model.predict_proba(X_test_scaled)[:, 1]

print(y_prob[:10])

## 11. ROC & AUC

In [ ]:
from sklearn.metrics import roc_curve, roc_auc_score

fpr, tpr, thresholds = roc_curve(y_test, y_prob)

print("FPR:", fpr[:5])
print("TPR:", tpr[:5])

In [ ]:
auc_score = roc_auc_score(y_test, y_prob)

print(f"AUC Score: {auc_score:.4f}")

In [ ]:
plt.figure(figsize=(7, 6))

plt.plot(
    fpr,
    tpr,
    label=f"Logistic Regression (AUC = {auc_score:.4f})"
)

plt.plot(
    [0, 1],
    [0, 1],
    linestyle="--",
    label="Random Classifier"
)

plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("ROC Curve")
plt.legend()
plt.grid()

plt.show()

## 12. Random Forest

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

rf_model.fit(X_train, y_train)

print("Random Forest trained successfully!")

In [ ]:
rf_pred = rf_model.predict(X_test)

rf_prob = rf_model.predict_proba(X_test)[:, 1]

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

rf_accuracy = accuracy_score(y_test, rf_pred)
rf_precision = precision_score(y_test, rf_pred)
rf_recall = recall_score(y_test, rf_pred)
rf_f1 = f1_score(y_test, rf_pred)
rf_auc = roc_auc_score(y_test, rf_prob)

print(f"Accuracy : {rf_accuracy:.4f}")
print(f"Precision: {rf_precision:.4f}")
print(f"Recall   : {rf_recall:.4f}")
print(f"F1 Score : {rf_f1:.4f}")
print(f"AUC      : {rf_auc:.4f}")

## 13.Model Comparison

In [ ]:
comparison = pd.DataFrame({
    "Metric": ["Accuracy", "Precision", "Recall", "F1 Score", "AUC"],
    "Logistic Regression": [
        accuracy,
        precision,
        recall,
        f1,
        auc_score
    ],
    "Random Forest": [
        rf_accuracy,
        rf_precision,
        rf_recall,
        rf_f1,
        rf_auc
    ]
})

comparison

In [ ]:
from sklearn.metrics import roc_curve

# Logistic Regression ROC
lr_fpr, lr_tpr, _ = roc_curve(y_test, y_prob)

# Random Forest ROC
rf_fpr, rf_tpr, _ = roc_curve(y_test, rf_prob)

plt.figure(figsize=(8, 6))

plt.plot(
    lr_fpr,
    lr_tpr,
    label=f"Logistic Regression (AUC = {auc_score:.4f})"
)

plt.plot(
    rf_fpr,
    rf_tpr,
    label=f"Random Forest (AUC = {rf_auc:.4f})"
)

plt.plot(
    [0, 1],
    [0, 1],
    linestyle="--",
    label="Random Classifier"
)

plt.xlabel("False Positive Rate")
plt.ylabel("True Positive Rate")
plt.title("ROC Curve Comparison")
plt.legend()
plt.grid()

plt.show()

## 14.Feature Importance

In [ ]:
feature_importance = pd.DataFrame({
    "Feature": X.columns,
    "Importance": rf_model.feature_importances_
})

feature_importance = feature_importance.sort_values(
    by="Importance",
    ascending=False
)

feature_importance.head(10)

In [ ]:
plt.figure(figsize=(10, 6))

sns.barplot(
    data=feature_importance.head(10),
    x="Importance",
    y="Feature"
)

plt.title("Top 10 Feature Importances - Random Forest")
plt.xlabel("Importance")
plt.ylabel("Feature")

plt.show()

In [ ]:
from sklearn.metrics import classification_report

print("Logistic Regression Classification Report")
print(classification_report(
    y_test,
    y_pred,
    target_names=["Malignant", "Benign"]
))

## 15. Conclusion

In this project, a binary classification system was developed using the Breast Cancer Wisconsin dataset from scikit-learn. The dataset contains 569 samples and 30 numerical features.

Two machine learning models were trained and evaluated:

- Logistic Regression
- Random Forest

The models were evaluated using Accuracy, Precision, Recall, F1 Score, ROC Curve, and AUC.

Logistic Regression achieved:
- Accuracy: 98.25%
- Precision: 98.61%
- Recall: 98.61%
- F1 Score: 98.61%
- AUC: 99.54%

Random Forest achieved:
- Accuracy: 95.61%
- Precision: 95.89%
- Recall: 97.22%
- F1 Score: 96.55%
- AUC: 99.37%

The results demonstrate that both models were able to distinguish between the two target classes effectively on the held-out test set. ROC curves and AUC provided an additional evaluation of the models across different classification thresholds.

This project provided practical experience in data preprocessing, exploratory data analysis, feature scaling, classification, model evaluation, confusion matrices, ROC curves, AUC, and feature importance.